{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script process DNA and RNA  integration results and merge them together side by side. DNA and RNA integration results can be made by concatenating the summary files from the integration pipeline. It is written for Cervical project by Stuart zong in Feburary 2017."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import seaborn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import ttest_ind\n",
    "import scipy.stats as ss\n",
    "from matplotlib import style\n",
    "matplotlib.style.use('ggplot')\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['figure.figsize'] = (15, 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function process groupby object, combine all rows with the same patient/virus indexed entries into one row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def process_grp(grp):\n",
    "    _df_or_sr = grp.apply(lambda col: ';'.join(col.unique().tolist()))\n",
    "    _df = _df_or_sr.to_frame().T\n",
    "    assert _df.shape[0] == 1\n",
    "    return _df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(271, 7)\n",
      "(226, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>external</th>\n",
       "      <th>virus</th>\n",
       "      <th>Integration</th>\n",
       "      <th>human-breakpoint</th>\n",
       "      <th>gene</th>\n",
       "      <th>library</th>\n",
       "      <th>patient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HTMCP_126</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37236</td>\n",
       "      <td>HTMCP-03-06-02003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HTMCP_133</td>\n",
       "      <td>HPV_73</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_36310359</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A37243</td>\n",
       "      <td>HTMCP-03-06-02012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    external   virus Integration human-breakpoint        gene library  \\\n",
       "2  HTMCP_126  HPV_16          NO                -           -  A37236   \n",
       "6  HTMCP_133  HPV_73         YES      17_36310359  intergenic  A37243   \n",
       "\n",
       "             patient  \n",
       "2  HTMCP-03-06-02003  \n",
       "6  HTMCP-03-06-02012  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_integration = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/integration/82_patients/genome/genome_integration_summary.csv.added', sep='\\t')\n",
    "print(RNA_integration.shape)\n",
    "RNA_integration = RNA_integration[RNA_integration['virus'] != 'No-Hits']\n",
    "print(RNA_integration.shape)\n",
    "RNA_integration.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>virus</th>\n",
       "      <th>external</th>\n",
       "      <th>Integration</th>\n",
       "      <th>human-breakpoint</th>\n",
       "      <th>gene</th>\n",
       "      <th>library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTMCP-03-06-02003</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_126</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HTMCP-03-06-02012</td>\n",
       "      <td>HPV_34</td>\n",
       "      <td>HTMCP_133</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HTMCP-03-06-02012</td>\n",
       "      <td>HPV_73</td>\n",
       "      <td>HTMCP_133</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_36310359</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A37243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HTMCP-03-06-02036</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_142</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTMCP-03-06-02037</td>\n",
       "      <td>HPV_52</td>\n",
       "      <td>HTMCP_143</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HTMCP-03-06-02042</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_145</td>\n",
       "      <td>YES</td>\n",
       "      <td>1_151029494;1_151029496</td>\n",
       "      <td>CDC42SE1</td>\n",
       "      <td>A37255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HTMCP-03-06-02046</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_147</td>\n",
       "      <td>YES</td>\n",
       "      <td>19_1181971;19_1183502;19_1192727</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A37257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HTMCP-03-06-02054</td>\n",
       "      <td>HPV_52</td>\n",
       "      <td>HTMCP_149</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HTMCP-03-06-02058</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_151</td>\n",
       "      <td>YES</td>\n",
       "      <td>18_37408866;8_128319365;8_128323246</td>\n",
       "      <td>intergenic;DQ515899,DQ515898,LOC727677</td>\n",
       "      <td>A37261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HTMCP-03-06-02063</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_156</td>\n",
       "      <td>YES</td>\n",
       "      <td>2_141933994;2_142008771;22_42896541</td>\n",
       "      <td>LRP1B;intergenic</td>\n",
       "      <td>A37266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HTMCP-03-06-02097</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_173</td>\n",
       "      <td>YES</td>\n",
       "      <td>2_19136363;GL000220.1_147778;GL000220.1_156305...</td>\n",
       "      <td>intergenic;RNA5-8S5</td>\n",
       "      <td>A37283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HTMCP-03-06-02098</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_174</td>\n",
       "      <td>YES</td>\n",
       "      <td>1_236331988</td>\n",
       "      <td>GPR137B</td>\n",
       "      <td>A37284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HTMCP-03-06-02098</td>\n",
       "      <td>HPV_9</td>\n",
       "      <td>HTMCP_174</td>\n",
       "      <td>YES</td>\n",
       "      <td>2_137648071</td>\n",
       "      <td>THSD7B</td>\n",
       "      <td>A37284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HTMCP-03-06-02103</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_230</td>\n",
       "      <td>YES</td>\n",
       "      <td>1_193547620;15_71989557</td>\n",
       "      <td>intergenic;THSD4</td>\n",
       "      <td>A38427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HTMCP-03-06-02103</td>\n",
       "      <td>HPV_53</td>\n",
       "      <td>HTMCP_230</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A38427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HTMCP-03-06-02108</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_233</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_78594577;17_78604388;17_78628159;17_7862875...</td>\n",
       "      <td>RPTOR;intergenic;FAM157B</td>\n",
       "      <td>A38430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HTMCP-03-06-02109</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_153</td>\n",
       "      <td>YES</td>\n",
       "      <td>11_123438967;1_249239730;12_52469492;12_524892...</td>\n",
       "      <td>GRAMD1B;intergenic;C12orf44;bpl;MGAT4C;ERBB2;I...</td>\n",
       "      <td>A37263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HTMCP-03-06-02110</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_235</td>\n",
       "      <td>YES</td>\n",
       "      <td>8_128296360;8_128297161;8_128300583;8_128316891</td>\n",
       "      <td>intergenic;DQ515899,DQ515898,LOC727677</td>\n",
       "      <td>A38432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HTMCP-03-06-02110</td>\n",
       "      <td>HPV_45</td>\n",
       "      <td>HTMCP_235</td>\n",
       "      <td>YES</td>\n",
       "      <td>8_128302005</td>\n",
       "      <td>DQ515899,DQ515898,LOC727677</td>\n",
       "      <td>A38432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HTMCP-03-06-02120</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_240</td>\n",
       "      <td>YES</td>\n",
       "      <td>GL000220.1_159017</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A38437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HTMCP-03-06-02128</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_244</td>\n",
       "      <td>YES</td>\n",
       "      <td>15_98072962;15_98072968;15_98076132;15_9807849...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A38441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HTMCP-03-06-02144</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_248</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A38445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HTMCP-03-06-02145</td>\n",
       "      <td>HPV_33</td>\n",
       "      <td>HTMCP_249</td>\n",
       "      <td>YES</td>\n",
       "      <td>13_41487747;13_41543293;13_41553397</td>\n",
       "      <td>SUGT1P3,TPTE2P5;ELF1</td>\n",
       "      <td>A38446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HTMCP-03-06-02145</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_249</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A38446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HTMCP-03-06-02145</td>\n",
       "      <td>HPV_9</td>\n",
       "      <td>HTMCP_249</td>\n",
       "      <td>YES</td>\n",
       "      <td>18_28136581</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A38446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HTMCP-03-06-02147</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_250</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_51244128;17_51286372;6_43405929;6_43565313;...</td>\n",
       "      <td>intergenic;ABCC10;POLH;IMPG1;BC012753;RNA5-8S5</td>\n",
       "      <td>A38447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HTMCP-03-06-02149</td>\n",
       "      <td>HPV_51</td>\n",
       "      <td>HTMCP_276</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HTMCP-03-06-02149</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_276</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HTMCP-03-06-02149</td>\n",
       "      <td>HPV_82</td>\n",
       "      <td>HTMCP_276</td>\n",
       "      <td>YES</td>\n",
       "      <td>8_128678796;8_128696014;8_128738871;8_12874469...</td>\n",
       "      <td>intergenic;BC042052</td>\n",
       "      <td>A52317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HTMCP-03-06-02151</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_278</td>\n",
       "      <td>YES</td>\n",
       "      <td>14_68703151;14_68703154</td>\n",
       "      <td>Metazoa_SRP,RAD51B</td>\n",
       "      <td>A54377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HTMCP-03-06-02155</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_280</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HTMCP-03-06-02156</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_281</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HTMCP-03-06-02156</td>\n",
       "      <td>HPV_33</td>\n",
       "      <td>HTMCP_281</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HTMCP-03-06-02156</td>\n",
       "      <td>HPV_52</td>\n",
       "      <td>HTMCP_281</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HTMCP-03-06-02156</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_281</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A52320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HTMCP-03-06-02174</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_286</td>\n",
       "      <td>YES</td>\n",
       "      <td>3_194761611;3_194766459;5_150329825;5_15035456...</td>\n",
       "      <td>intergenic;SEMA3D</td>\n",
       "      <td>A52323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HTMCP-03-06-02175</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_287</td>\n",
       "      <td>YES</td>\n",
       "      <td>10_13326755;10_13389178;10_88649321;1_17268995...</td>\n",
       "      <td>PHYH;SEPHS1;BMPR1A;CROCC;intergenic;PAK1;DLG2;...</td>\n",
       "      <td>A52324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HTMCP-03-06-02176</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_288</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_41388327;17_41392842;17_41466353;17_7917951...</td>\n",
       "      <td>intergenic;AZI1;MAML3</td>\n",
       "      <td>A52325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HTMCP-03-06-02176</td>\n",
       "      <td>HPV_58</td>\n",
       "      <td>HTMCP_288</td>\n",
       "      <td>YES</td>\n",
       "      <td>4_140811717</td>\n",
       "      <td>MAML3</td>\n",
       "      <td>A52325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HTMCP-03-06-02179</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_290</td>\n",
       "      <td>YES</td>\n",
       "      <td>17_80649353;6_3260628;6_3260654;6_3261374;6_32...</td>\n",
       "      <td>RAB40B;VARS,CFB,PSMG4,AK096219,SKIV2L,C4B_2,C2...</td>\n",
       "      <td>A54381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HTMCP-03-06-02179</td>\n",
       "      <td>HPV_97</td>\n",
       "      <td>HTMCP_290</td>\n",
       "      <td>YES</td>\n",
       "      <td>6_3289268</td>\n",
       "      <td>SLC22A23,CFB,TNXB,C4B_2,C2,HSPA1L</td>\n",
       "      <td>A54381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HTMCP-03-06-02180</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_291</td>\n",
       "      <td>YES</td>\n",
       "      <td>14_68436523;2_147596973</td>\n",
       "      <td>RAD51B;intergenic</td>\n",
       "      <td>A52327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HTMCP-03-06-02185</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_292</td>\n",
       "      <td>YES</td>\n",
       "      <td>2_43630082;2_43674215</td>\n",
       "      <td>THADA</td>\n",
       "      <td>A52328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>HTMCP-03-06-02195</td>\n",
       "      <td>HPV_45</td>\n",
       "      <td>HTMCP_399</td>\n",
       "      <td>YES</td>\n",
       "      <td>6_96414963</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A57084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>HTMCP-03-06-02202</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_400</td>\n",
       "      <td>YES</td>\n",
       "      <td>14_99656850;14_99703940;14_99711127;14_9971123...</td>\n",
       "      <td>BCL11B</td>\n",
       "      <td>A57085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>HTMCP-03-06-02205</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_304</td>\n",
       "      <td>YES</td>\n",
       "      <td>3_59700106</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A54385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>HTMCP-03-06-02210</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_307</td>\n",
       "      <td>YES</td>\n",
       "      <td>10_99063964</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A52339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>HTMCP-03-06-02213</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_309</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A54387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>HTMCP-03-06-02214</td>\n",
       "      <td>HPV_16</td>\n",
       "      <td>HTMCP_310</td>\n",
       "      <td>YES</td>\n",
       "      <td>4_87540666;4_87644435</td>\n",
       "      <td>PTPN13</td>\n",
       "      <td>A52340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>HTMCP-03-06-02225</td>\n",
       "      <td>HPV_69</td>\n",
       "      <td>HTMCP_406</td>\n",
       "      <td>YES</td>\n",
       "      <td>6_127148053</td>\n",
       "      <td>AK127472</td>\n",
       "      <td>A57091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>HTMCP-03-06-02238</td>\n",
       "      <td>HPV_18</td>\n",
       "      <td>HTMCP_409</td>\n",
       "      <td>YES</td>\n",
       "      <td>15_72233495;15_72236104;16_79363861;16_7942275...</td>\n",
       "      <td>MYO9A;intergenic;COL6A3;LRRFIP1</td>\n",
       "      <td>A57094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              patient   virus   external Integration  \\\n",
       "0   HTMCP-03-06-02003  HPV_16  HTMCP_126          NO   \n",
       "1   HTMCP-03-06-02012  HPV_34  HTMCP_133          NO   \n",
       "2   HTMCP-03-06-02012  HPV_73  HTMCP_133         YES   \n",
       "3   HTMCP-03-06-02036  HPV_16  HTMCP_142          NO   \n",
       "4   HTMCP-03-06-02037  HPV_52  HTMCP_143          NO   \n",
       "5   HTMCP-03-06-02042  HPV_16  HTMCP_145         YES   \n",
       "6   HTMCP-03-06-02046  HPV_16  HTMCP_147         YES   \n",
       "7   HTMCP-03-06-02054  HPV_52  HTMCP_149          NO   \n",
       "8   HTMCP-03-06-02058  HPV_18  HTMCP_151         YES   \n",
       "9   HTMCP-03-06-02063  HPV_16  HTMCP_156         YES   \n",
       "10  HTMCP-03-06-02097  HPV_16  HTMCP_173         YES   \n",
       "11  HTMCP-03-06-02098  HPV_18  HTMCP_174         YES   \n",
       "12  HTMCP-03-06-02098   HPV_9  HTMCP_174         YES   \n",
       "13  HTMCP-03-06-02103  HPV_18  HTMCP_230         YES   \n",
       "14  HTMCP-03-06-02103  HPV_53  HTMCP_230          NO   \n",
       "15  HTMCP-03-06-02108  HPV_16  HTMCP_233         YES   \n",
       "16  HTMCP-03-06-02109  HPV_16  HTMCP_153         YES   \n",
       "17  HTMCP-03-06-02110  HPV_18  HTMCP_235         YES   \n",
       "18  HTMCP-03-06-02110  HPV_45  HTMCP_235         YES   \n",
       "19  HTMCP-03-06-02120  HPV_58  HTMCP_240         YES   \n",
       "20  HTMCP-03-06-02128  HPV_16  HTMCP_244         YES   \n",
       "21  HTMCP-03-06-02144  HPV_16  HTMCP_248          NO   \n",
       "22  HTMCP-03-06-02145  HPV_33  HTMCP_249         YES   \n",
       "23  HTMCP-03-06-02145  HPV_58  HTMCP_249          NO   \n",
       "24  HTMCP-03-06-02145   HPV_9  HTMCP_249         YES   \n",
       "25  HTMCP-03-06-02147  HPV_16  HTMCP_250         YES   \n",
       "26  HTMCP-03-06-02149  HPV_51  HTMCP_276          NO   \n",
       "27  HTMCP-03-06-02149  HPV_58  HTMCP_276          NO   \n",
       "28  HTMCP-03-06-02149  HPV_82  HTMCP_276         YES   \n",
       "29  HTMCP-03-06-02151  HPV_16  HTMCP_278         YES   \n",
       "30  HTMCP-03-06-02155  HPV_16  HTMCP_280          NO   \n",
       "31  HTMCP-03-06-02156  HPV_16  HTMCP_281          NO   \n",
       "32  HTMCP-03-06-02156  HPV_33  HTMCP_281          NO   \n",
       "33  HTMCP-03-06-02156  HPV_52  HTMCP_281          NO   \n",
       "34  HTMCP-03-06-02156  HPV_58  HTMCP_281          NO   \n",
       "35  HTMCP-03-06-02174  HPV_16  HTMCP_286         YES   \n",
       "36  HTMCP-03-06-02175  HPV_58  HTMCP_287         YES   \n",
       "37  HTMCP-03-06-02176  HPV_16  HTMCP_288         YES   \n",
       "38  HTMCP-03-06-02176  HPV_58  HTMCP_288         YES   \n",
       "39  HTMCP-03-06-02179  HPV_18  HTMCP_290         YES   \n",
       "40  HTMCP-03-06-02179  HPV_97  HTMCP_290         YES   \n",
       "41  HTMCP-03-06-02180  HPV_18  HTMCP_291         YES   \n",
       "42  HTMCP-03-06-02185  HPV_18  HTMCP_292         YES   \n",
       "43  HTMCP-03-06-02195  HPV_45  HTMCP_399         YES   \n",
       "44  HTMCP-03-06-02202  HPV_16  HTMCP_400         YES   \n",
       "45  HTMCP-03-06-02205  HPV_16  HTMCP_304         YES   \n",
       "46  HTMCP-03-06-02210  HPV_16  HTMCP_307         YES   \n",
       "47  HTMCP-03-06-02213  HPV_16  HTMCP_309          NO   \n",
       "48  HTMCP-03-06-02214  HPV_16  HTMCP_310         YES   \n",
       "49  HTMCP-03-06-02225  HPV_69  HTMCP_406         YES   \n",
       "50  HTMCP-03-06-02238  HPV_18  HTMCP_409         YES   \n",
       "\n",
       "                                     human-breakpoint  \\\n",
       "0                                                   -   \n",
       "1                                                   -   \n",
       "2                                         17_36310359   \n",
       "3                                                   -   \n",
       "4                                                   -   \n",
       "5                             1_151029494;1_151029496   \n",
       "6                    19_1181971;19_1183502;19_1192727   \n",
       "7                                                   -   \n",
       "8                 18_37408866;8_128319365;8_128323246   \n",
       "9                 2_141933994;2_142008771;22_42896541   \n",
       "10  2_19136363;GL000220.1_147778;GL000220.1_156305...   \n",
       "11                                        1_236331988   \n",
       "12                                        2_137648071   \n",
       "13                            1_193547620;15_71989557   \n",
       "14                                                  -   \n",
       "15  17_78594577;17_78604388;17_78628159;17_7862875...   \n",
       "16  11_123438967;1_249239730;12_52469492;12_524892...   \n",
       "17    8_128296360;8_128297161;8_128300583;8_128316891   \n",
       "18                                        8_128302005   \n",
       "19                                  GL000220.1_159017   \n",
       "20  15_98072962;15_98072968;15_98076132;15_9807849...   \n",
       "21                                                  -   \n",
       "22                13_41487747;13_41543293;13_41553397   \n",
       "23                                                  -   \n",
       "24                                        18_28136581   \n",
       "25  17_51244128;17_51286372;6_43405929;6_43565313;...   \n",
       "26                                                  -   \n",
       "27                                                  -   \n",
       "28  8_128678796;8_128696014;8_128738871;8_12874469...   \n",
       "29                            14_68703151;14_68703154   \n",
       "30                                                  -   \n",
       "31                                                  -   \n",
       "32                                                  -   \n",
       "33                                                  -   \n",
       "34                                                  -   \n",
       "35  3_194761611;3_194766459;5_150329825;5_15035456...   \n",
       "36  10_13326755;10_13389178;10_88649321;1_17268995...   \n",
       "37  17_41388327;17_41392842;17_41466353;17_7917951...   \n",
       "38                                        4_140811717   \n",
       "39  17_80649353;6_3260628;6_3260654;6_3261374;6_32...   \n",
       "40                                          6_3289268   \n",
       "41                            14_68436523;2_147596973   \n",
       "42                              2_43630082;2_43674215   \n",
       "43                                         6_96414963   \n",
       "44  14_99656850;14_99703940;14_99711127;14_9971123...   \n",
       "45                                         3_59700106   \n",
       "46                                        10_99063964   \n",
       "47                                                  -   \n",
       "48                              4_87540666;4_87644435   \n",
       "49                                        6_127148053   \n",
       "50  15_72233495;15_72236104;16_79363861;16_7942275...   \n",
       "\n",
       "                                                 gene library  \n",
       "0                                                   -  A37236  \n",
       "1                                                   -  A37243  \n",
       "2                                          intergenic  A37243  \n",
       "3                                                   -  A37252  \n",
       "4                                                   -  A37253  \n",
       "5                                            CDC42SE1  A37255  \n",
       "6                                          intergenic  A37257  \n",
       "7                                                   -  A37259  \n",
       "8              intergenic;DQ515899,DQ515898,LOC727677  A37261  \n",
       "9                                    LRP1B;intergenic  A37266  \n",
       "10                                intergenic;RNA5-8S5  A37283  \n",
       "11                                            GPR137B  A37284  \n",
       "12                                             THSD7B  A37284  \n",
       "13                                   intergenic;THSD4  A38427  \n",
       "14                                                  -  A38427  \n",
       "15                           RPTOR;intergenic;FAM157B  A38430  \n",
       "16  GRAMD1B;intergenic;C12orf44;bpl;MGAT4C;ERBB2;I...  A37263  \n",
       "17             intergenic;DQ515899,DQ515898,LOC727677  A38432  \n",
       "18                        DQ515899,DQ515898,LOC727677  A38432  \n",
       "19                                         intergenic  A38437  \n",
       "20                                         intergenic  A38441  \n",
       "21                                                  -  A38445  \n",
       "22                               SUGT1P3,TPTE2P5;ELF1  A38446  \n",
       "23                                                  -  A38446  \n",
       "24                                         intergenic  A38446  \n",
       "25     intergenic;ABCC10;POLH;IMPG1;BC012753;RNA5-8S5  A38447  \n",
       "26                                                  -  A52317  \n",
       "27                                                  -  A52317  \n",
       "28                                intergenic;BC042052  A52317  \n",
       "29                                 Metazoa_SRP,RAD51B  A54377  \n",
       "30                                                  -  A52319  \n",
       "31                                                  -  A52320  \n",
       "32                                                  -  A52320  \n",
       "33                                                  -  A52320  \n",
       "34                                                  -  A52320  \n",
       "35                                  intergenic;SEMA3D  A52323  \n",
       "36  PHYH;SEPHS1;BMPR1A;CROCC;intergenic;PAK1;DLG2;...  A52324  \n",
       "37                              intergenic;AZI1;MAML3  A52325  \n",
       "38                                              MAML3  A52325  \n",
       "39  RAB40B;VARS,CFB,PSMG4,AK096219,SKIV2L,C4B_2,C2...  A54381  \n",
       "40                  SLC22A23,CFB,TNXB,C4B_2,C2,HSPA1L  A54381  \n",
       "41                                  RAD51B;intergenic  A52327  \n",
       "42                                              THADA  A52328  \n",
       "43                                         intergenic  A57084  \n",
       "44                                             BCL11B  A57085  \n",
       "45                                         intergenic  A54385  \n",
       "46                                         intergenic  A52339  \n",
       "47                                                  -  A54387  \n",
       "48                                             PTPN13  A52340  \n",
       "49                                           AK127472  A57091  \n",
       "50                    MYO9A;intergenic;COL6A3;LRRFIP1  A57094  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_GRP = RNA_integration.groupby(['patient', 'virus']).apply(process_grp).reset_index(level=2, drop=True)\n",
    "del RNA_GRP['patient']\n",
    "del RNA_GRP['virus']\n",
    "# RNA_GRP\n",
    "RNA_GRP.reset_index(inplace=True)\n",
    "RNA_GRP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# RNA_GRP['patient'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(520, 6)\n",
      "(513, 6)\n"
     ]
    }
   ],
   "source": [
    "DNA_integration = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/integration/82_patients/transcriptome/transcriptome_integration_summary.csv.added', sep='\\t')\n",
    "print(DNA_integration.shape)\n",
    "DNA_integration = DNA_integration[DNA_integration['virus'] != 'No-Hits']\n",
    "print(DNA_integration.shape)\n",
    "\n",
    "DNA_GRP = DNA_integration.groupby(['patient', 'virus']).apply(process_grp).reset_index(level=2, drop=True)\n",
    "del DNA_GRP['patient']\n",
    "del DNA_GRP['virus']\n",
    "# DNA_GRP\n",
    "DNA_GRP.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge DNA and RNA integration results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "merged = pd.merge(DNA_GRP, RNA_GRP, on=['patient', 'virus'], how='outer', suffixes=('_RNA', '_DNA'))\n",
    "merged.fillna('na', inplace=True)\n",
    "merged.to_csv('test_final_merged.csv', sep='\\t')\n",
    "merged.head()\n",
    "\n",
    "np.size(merged.patient.unique())\n",
    "strains = merged.virus.unique()\n",
    "strains\n",
    "\n",
    "\n",
    "meta = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/variant_bwamem/82_patients/HIV_Cervical_82_pateints_vcfs_bams.csv', sep='\\t')\n",
    "all_patients = meta.patient.unique()\n",
    "np.size(all_patients)\n",
    "merged.set_index(['patient', 'virus'], inplace=True)\n",
    "\n",
    "df_patients = list(set([i[0] for i in merged.index.unique().tolist()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>library_RNA</th>\n",
       "      <th>Integration_RNA</th>\n",
       "      <th>human-breakpoint_RNA</th>\n",
       "      <th>gene_RNA</th>\n",
       "      <th>external</th>\n",
       "      <th>Integration_DNA</th>\n",
       "      <th>human-breakpoint_DNA</th>\n",
       "      <th>gene_DNA</th>\n",
       "      <th>library_DNA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient</th>\n",
       "      <th>virus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HTMCP-03-06-02001</th>\n",
       "      <th>HPV_16</th>\n",
       "      <td>A37700</td>\n",
       "      <td>YES</td>\n",
       "      <td>3_189526056;3_189526059;3_189526315;3_18952631...</td>\n",
       "      <td>TP63</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTMCP-03-06-02002</th>\n",
       "      <th>HPV_52</th>\n",
       "      <td>A37701</td>\n",
       "      <td>YES</td>\n",
       "      <td>22_43161525;22_43161526;22_43163674;22_4316513...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTMCP-03-06-02003</th>\n",
       "      <th>HPV_16</th>\n",
       "      <td>A37702</td>\n",
       "      <td>YES</td>\n",
       "      <td>11_60882522;11_60935033</td>\n",
       "      <td>CD5;intergenic</td>\n",
       "      <td>HTMCP_126</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A37236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HTMCP-03-06-02006</th>\n",
       "      <th>HPV_18</th>\n",
       "      <td>A37704</td>\n",
       "      <td>NO</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HPV_45</th>\n",
       "      <td>A37704</td>\n",
       "      <td>YES</td>\n",
       "      <td>12_46287201;12_46328672;12_46329948;12_4633703...</td>\n",
       "      <td>ARID2;SCAF11</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         library_RNA Integration_RNA  \\\n",
       "patient           virus                                \n",
       "HTMCP-03-06-02001 HPV_16      A37700             YES   \n",
       "HTMCP-03-06-02002 HPV_52      A37701             YES   \n",
       "HTMCP-03-06-02003 HPV_16      A37702             YES   \n",
       "HTMCP-03-06-02006 HPV_18      A37704              NO   \n",
       "                  HPV_45      A37704             YES   \n",
       "\n",
       "                                                       human-breakpoint_RNA  \\\n",
       "patient           virus                                                       \n",
       "HTMCP-03-06-02001 HPV_16  3_189526056;3_189526059;3_189526315;3_18952631...   \n",
       "HTMCP-03-06-02002 HPV_52  22_43161525;22_43161526;22_43163674;22_4316513...   \n",
       "HTMCP-03-06-02003 HPV_16                            11_60882522;11_60935033   \n",
       "HTMCP-03-06-02006 HPV_18                                                  -   \n",
       "                  HPV_45  12_46287201;12_46328672;12_46329948;12_4633703...   \n",
       "\n",
       "                                gene_RNA   external Integration_DNA  \\\n",
       "patient           virus                                               \n",
       "HTMCP-03-06-02001 HPV_16            TP63         na              na   \n",
       "HTMCP-03-06-02002 HPV_52      intergenic         na              na   \n",
       "HTMCP-03-06-02003 HPV_16  CD5;intergenic  HTMCP_126              NO   \n",
       "HTMCP-03-06-02006 HPV_18               -         na              na   \n",
       "                  HPV_45    ARID2;SCAF11         na              na   \n",
       "\n",
       "                         human-breakpoint_DNA gene_DNA library_DNA  \n",
       "patient           virus                                             \n",
       "HTMCP-03-06-02001 HPV_16                   na       na          na  \n",
       "HTMCP-03-06-02002 HPV_52                   na       na          na  \n",
       "HTMCP-03-06-02003 HPV_16                    -        -      A37236  \n",
       "HTMCP-03-06-02006 HPV_18                   na       na          na  \n",
       "                  HPV_45                   na       na          na  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "find all genes hit by integration\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['FHIT', 'STK11', 'NELFE', 'THADA', 'LYRM7', 'BC066991', 'RREB1', 'DLG2', 'PKD1L1', 'MIB1', 'UPK1B', 'CDC42SE1', 'PRIM2', 'MYC', 'CDH5', 'ERBB2', 'KRT32', 'RNF103-CHMP3', 'KRT26', 'LOC440434', 'CT64', 'KCNN3', 'AK127472', 'C12orf44', 'PTPN13', 'bpl', 'ACSL5', 'DQ515899', 'DQ515898', 'RNF103', 'RNA5-8S5', 'BC042052', 'RASSF6', 'CRTC3', 'CLTA', 'GPR137B', 'FNBP1L', 'MAML3', 'PSMG4', 'SKIV2L', 'RASEF', 'CEACAM6', 'TUBD1', 'B4GALT4', 'ARID2', 'LY86-AS1', 'MLLT11', 'COL4A5', 'GUCY2GP', 'RAB40B', 'ELOVL2', 'CYP21A1P', 'KIAA1958', 'SMIM19', 'BC037497', 'CD5', 'SERHL2', 'RMND5A', 'HINT1', 'TP63', 'HV975509', 'TBC1D3', 'BEAN1', 'C4B_2', 'C2', 'PRUNE', 'SUGT1P3', 'FGF12', 'RAD51B', 'IRS4', 'intergenic', 'IRS1', 'TPTE2P5', 'VARS', 'LOC646762', 'ELF1', 'FAM83B', 'MIEN1', 'LOC727677', 'KLF12', 'LRP1B', 'SCAF11', 'CYP21A2', 'NR4A1', 'TMEM242', 'AK096219', 'SLC20A2', 'MPPED2', 'CFB', 'CLPTM1L', 'ANKFN1', 'WBP4']\n"
     ]
    }
   ],
   "source": [
    "rna_genes = merged['gene_RNA'].tolist()\n",
    "rna_genes = [i.split(';') for i in rna_genes]\n",
    "rna_genes = list(set([item for sublist in rna_genes for item in sublist]))\n",
    "rna_genes = [i.split(',') for i in rna_genes]\n",
    "# print(rna_genes)\n",
    "\n",
    "rna_genes = list(set([item for sublist in rna_genes for item in sublist]))\n",
    "# print(rna_genes)\n",
    "\n",
    "rna_genes.remove('-')\n",
    "rna_genes.remove('na')\n",
    "print(rna_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "figure out which patients have no integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A37705', 'A37711', 'A37714', 'A37724', 'A37733', 'A50336', 'A50337', 'A50341', 'A50343', 'A50344', 'A54469', 'A54470', 'A52346', 'A52349', 'A57109', 'A57112', 'A57113', 'A57115', 'A57121', 'A57122']\n"
     ]
    }
   ],
   "source": [
    "lib_all = meta['RNA_lib'].dropna().tolist()\n",
    "# lib_all\n",
    "libs_intg = merged.reset_index().query('Integration_RNA == \"{0}\"'.format('YES'))['library_RNA'].unique().tolist()\n",
    "libs_unintg = [item for item in lib_all if item not in libs_intg]\n",
    "print(libs_unintg)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge 2 dictionaries and create a new data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "x = {k: 'integrated' for k in libs_intg}\n",
    "y = {k: 'unintegrated' for k in libs_unintg}\n",
    "z = x.copy()\n",
    "z.update(y)\n",
    "z\n",
    "intgrs = pd.DataFrame(z.items(), columns=['library_RNA', 'integration_status'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "intgrs.to_csv('integration_status.txt', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merge_intg = merged.merge(intgrs, on='library_RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "merge_intg.to_csv('merged_integration_results.txt', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "add entries for all patients and strains for those absent from the summary file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for patient in all_patients:\n",
    "    df_strains = []\n",
    "    if patient in df_patients:\n",
    "        df_strains = merged.loc[(patient)].index.unique().tolist()\n",
    "        for strain in strains:\n",
    "            if strain not in df_strains:\n",
    "                merged.loc[(patient, strain),:] = pd.Series({i:'bbbb' for i in merged.columns}) \n",
    "    else:\n",
    "        for strain in strains:\n",
    "            merged.loc[(patient, strain),:] = pd.Series({i:'bbbb' for i in merged.columns})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged.to_csv('final_final_test.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# below are experiment code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# need to sort by virus then patient\n",
    "merged.index = merged.index.swaplevel()\n",
    "\n",
    "merged = merged.sortlevel()\n",
    "\n",
    "# merged.index = merged.index.swaplevel()\n",
    "\n",
    "# merged.head(200)\n",
    "# seems not working, use the python script to format RNA and DNA integration results to plot separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Human papillomavirus type 51',\n",
       " 'Human papillomavirus type 69',\n",
       " 'Human papillomavirus type 82']"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.loc[('HTMCP-03-06-02225')].index.get_level_values('virus').unique().tolist()\n",
    "# multiindex is a tuple\n",
    "merged.loc[('HTMCP-03-06-02225','Human papillomavirus type 69')]\n",
    "# list comprehension to make a dictionary and then to series\n",
    "pd.Series({i:'bbbb' for i in merged.columns})\n",
    "# add a entry for multiindexed data frame\n",
    "merged.loc[('HTMCP-03-06-02225', 'Human papillomavirus type 999'),:] = pd.Series({i:'not_detected' for i in merged.columns})\n",
    "\n",
    "# add all negative patients\n",
    "df_patients = list(set([i[0] for i in merged.index.unique().tolist()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "patient            virus \n",
       "HTMCP-03-06-02001  HPV_16    YES\n",
       "HTMCP-03-06-02002  HPV_52    YES\n",
       "HTMCP-03-06-02003  HPV_16    YES\n",
       "HTMCP-03-06-02006  HPV_18     NO\n",
       "                   HPV_45    YES\n",
       "Name: Integration_RNA, dtype: object"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged['virus', 'patient', 'Integration_RNA']\n",
    "merged['Integration_RNA'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HTMCP-03-06-02113', 'HTMCP-03-06-02036']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/variant_bwamem/82_patients/HIV_Cervical_82_pateints_vcfs_bams.csv', sep='\\t')\n",
    "all_patients = meta.patient.unique()\n",
    "np.size(all_patients)\n",
    "# all_patients\n",
    "df_patients = list(set([i[0] for i in merged.index.unique().tolist()]))\n",
    "df_patients[:2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "combine dna integration results into one patient per line format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def process_grp(grp):\n",
    "    _df_or_sr = grp.apply(lambda col: ';'.join(col.unique().tolist()))\n",
    "    _df = _df_or_sr.to_frame().T\n",
    "    assert _df.shape[0] == 1\n",
    "    return _df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(271, 7)\n",
      "(226, 7)\n"
     ]
    }
   ],
   "source": [
    "DNA_integration = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/integration/82_patients/genome/genome_integration_summary.csv.added', sep='\\t')\n",
    "print(DNA_integration.shape)\n",
    "DNA_integration = DNA_integration[DNA_integration['virus'] != 'No-Hits']\n",
    "print(DNA_integration.shape)\n",
    "# DNA_integration\n",
    "\n",
    "DNA_GRP = DNA_integration.groupby(['patient', 'virus']).apply(process_grp).reset_index(level=2, drop=True)\n",
    "del DNA_GRP['patient']\n",
    "del DNA_GRP['virus']\n",
    "# DNA_GRP\n",
    "DNA_GRP.reset_index(inplace=True)\n",
    "# DNA_GRP\n",
    "\n",
    "# this gives all the hpy strains (integrated or un-integrated) by patient\n",
    "DNA_GRP = DNA_GRP.groupby('patient').apply(process_grp).reset_index(drop=True).sort_values('patient')\n",
    "\n",
    "DNA_GRP = DNA_GRP.set_index('patient')\n",
    "idx = DNA_GRP.index\n",
    "\n",
    "DNA_GRP.head(2)\n",
    "\n",
    "for patient in all_patients:\n",
    "    if patient not in idx:\n",
    "#         print('no')\n",
    "        DNA_GRP.loc[patient] = pd.Series({i:'NO' for i in DNA_GRP.columns})\n",
    "\n",
    "DNA_GRP.reset_index().sort_values('patient')[['patient', 'virus', 'Integration']].to_csv('integration_results_dna_contig_level.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(520, 6)\n",
      "(513, 6)\n"
     ]
    }
   ],
   "source": [
    "RNA_integration = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/integration/82_patients/transcriptome/transcriptome_integration_summary.csv.added', sep='\\t')\n",
    "print(RNA_integration.shape)\n",
    "RNA_integration = RNA_integration[RNA_integration['virus'] != 'No-Hits']\n",
    "print(RNA_integration.shape)\n",
    "# RNA_integration\n",
    "\n",
    "RNA_GRP = RNA_integration.groupby(['patient', 'virus']).apply(process_grp).reset_index(level=2, drop=True)\n",
    "del RNA_GRP['patient']\n",
    "del RNA_GRP['virus']\n",
    "# RNA_GRP\n",
    "RNA_GRP.reset_index(inplace=True)\n",
    "# RNA_GRP\n",
    "\n",
    "# this gives all the hpy strains (integrated or un-integrated) by patient\n",
    "RNA_GRP = RNA_GRP.groupby('patient').apply(process_grp).reset_index(drop=True).sort_values('patient')\n",
    "\n",
    "RNA_GRP = RNA_GRP.set_index('patient')\n",
    "idx = RNA_GRP.index\n",
    "\n",
    "RNA_GRP.head(2)\n",
    "\n",
    "for patient in all_patients:\n",
    "    if patient not in idx:\n",
    "#         print('no')\n",
    "        RNA_GRP.loc[patient] = pd.Series({i:'NO' for i in RNA_GRP.columns})\n",
    "\n",
    "RNA_GRP.reset_index().sort_values('patient')[['patient', 'virus', 'Integration']].to_csv('integration_results_RNA_contig_level.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'virus', u'external', u'Integration', u'human-breakpoint', u'gene',\n",
       "       u'library'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
