{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### import tempfile\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "#import scipy.optimize as spo\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import style\n",
    "\n",
    "import collections\n",
    "import headers as HEADER\n",
    "import os, stat, os.path, time, datetime, subprocess\n",
    "import re, sys, glob, argparse, csv, shutil, fileinput\n",
    "from pprint import pprint\n",
    "from itertools import islice\n",
    "import operator\n",
    "import ConfigParser\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "\n",
    "from jinja2 import Environment, FileSystemLoader\n",
    "\n",
    "style.use('ggplot')\n",
    "\n",
    "\n",
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description='get coverage for a list of positions')\n",
    "    parser.add_argument(\n",
    "        '-m', '--meta_file',\n",
    "        help='specify input file which is meta file',\n",
    "        required=True)\n",
    "    parser.add_argument(\n",
    "        '-b', '--bed_regions',\n",
    "        help='specify region of interest in bed format',\n",
    "        required=True)\n",
    "    args = parser.parse_args()\n",
    "    return args\n",
    "\n",
    "\n",
    "def get_files(bam_vcf_files, input_headers):\n",
    "    \"\"\"Dictionary:{patient->{status->{file_identifier -> file_path}}}\"\"\"\n",
    "    patient_files = dict()\n",
    "    with open(bam_vcf_files, 'r') as fh:\n",
    "        records = csv.DictReader(fh,  delimiter='\\t')\n",
    "        headers = set(records.fieldnames)\n",
    "        input_headers = set(input_headers)\n",
    "        # check if all mandatory columns prsent in input file\n",
    "        if input_headers.issubset(headers):\n",
    "            print \"input file has all mandatory columns! Continue...\"\n",
    "            for line in records:\n",
    "                patient = line['patient']\n",
    "                status = line['status']\n",
    "                if patient not in patient_files:\n",
    "                    patient_files[patient] = {}\n",
    "                if status not in patient_files[patient]:\n",
    "                    patient_files[patient][status] = line\n",
    "                else:\n",
    "                    print 'Duplicate status entries! One tissue sequenced multiple times?'\n",
    "                    sys.exit()\n",
    "        else:\n",
    "            print \"Input file doesn't contain all mandatory headers!\\n %s\" % list(input_headers)\n",
    "            sys.exit()\n",
    "        return patient_files\n",
    "    \n",
    "    \n",
    "\n",
    "def make_BEDcoverage_script(patient_files, bed_regions, template_dir):\n",
    "    bed_path = \"/home/rcorbett/aligners/bedtools/BEDTools-Version-2.15.0/bin/\"\n",
    "    sam_path = \"/home/rcorbett/aligners/samtools/samtools-0.1.17/\"\n",
    "    wkdir = os.getcwd()\n",
    "    scripts = []\n",
    "    stamps = []\n",
    "    coverage_files = []\n",
    "    for patient in patient_files:\n",
    "        for status in patient_files[patient]:\n",
    "            library = patient_files[patient][status]['DNA_lib']\n",
    "            bam = patient_files[patient][status]['DNA_bam'] \n",
    "            complete_stamp = \".\".join([patient, status, library, \"complete\"])\n",
    "            stamps.append(complete_stamp)\n",
    "            coverage_file = \".\".join([patient, status, library, \"exon.coverage.txt\"])\n",
    "            coverage_files.append(coverage_file)\n",
    "            script = populate_template(patient,\n",
    "                                       status,\n",
    "                                       wkdir,\n",
    "                                       library,\n",
    "                                       bed_regions,\n",
    "                                       bam,\n",
    "                                       template_dir)\n",
    "            scripts.append(script)\n",
    "    return [scripts, stamps, coverage_files]\n",
    "    \n",
    "    \n",
    "def populate_template(patient, status, wkdir, library,\n",
    "                      bed_regions, bam, template_dir):\n",
    "    bedcoverage_script = \".\".join([patient, status, library, \"BEDcoverage.sh\"])\n",
    "    jinja2_env = Environment(loader=FileSystemLoader([template_dir]),\n",
    "                             trim_blocks=True)\n",
    "    template = jinja2_env.get_template('bedcoverage_template.sh')\n",
    "    with open(bedcoverage_script, 'wb') as opf:\n",
    "        content = template.render(patient=patient,\n",
    "                                  status=status,\n",
    "                                  library=library,\n",
    "                                  bam=bam,\n",
    "                                  bed_regions=bed_regions,\n",
    "                                  wkdir=wkdir)\n",
    "        opf.write(content)\n",
    "#         print 'templated {0}'.format(bedcoverage_script)\n",
    "    return bedcoverage_script\n",
    "\n",
    "\n",
    "# def make_dir(path):\n",
    "#     if not os.path.exists(path):\n",
    "#         os.makedirs(path)\n",
    "#     else:\n",
    "#         print \"%s directory already exists!\" % path\n",
    "      \n",
    "\n",
    "\n",
    "def intervals_to_dict(gene_intervals_file):\n",
    "    \"\"\"Dictionary holds {gene-> [chromosome, start, end]}\"\"\"\n",
    "    with open(gene_intervals_file, 'r') as fh:\n",
    "        records = csv.DictReader(fh,  delimiter='\\t')\n",
    "        gene_regions = dict()\n",
    "        for line in records:\n",
    "            gene = line['gene'].split('_')[0]\n",
    "            chromosome = line['chr']\n",
    "            start = line['start']\n",
    "            end = line['end']\n",
    "            info = [chromosome, start, end]\n",
    "            try:\n",
    "                gene_regions[gene].append(info)\n",
    "            except KeyError:\n",
    "                gene_regions[gene] = [info]\n",
    "    return gene_regions\n",
    "            \n",
    "           \n",
    "def  intervals_to_regions(gene_regions):\n",
    "    \"\"\"Calculate bedcoverage region for a gene including 1 kb up and downstream\"\"\"\n",
    "    for gene in  gene_regions:\n",
    "        #make_dir(gene)\n",
    "        #region_file = os.getcwd() + \"/\" + gene + \"/\" + gene + \"_combined_exon_pm_1kb.txt\"\n",
    "        region_file = os.getcwd() + \"/\" + gene + \"_combined_exon_pm_1kb.txt\"\n",
    "        starts = [int(i[1]) for i in gene_regions[gene]]\n",
    "        ends = [int(i[2]) for i in gene_regions[gene]]\n",
    "        chromosome = gene_regions[gene][0][0]\n",
    "        # expand region to 1 kb up and downstream the gene\n",
    "        interested_start = np.array(starts).min() - 1000\n",
    "        interested_end = np.array(ends).max() + 1000\n",
    "        with open(region_file, 'w') as opf:\n",
    "            # csv writer always generate a dos file\n",
    "            writer = csv.writer(opf,  delimiter='\\t', lineterminator=\"\\n\")\n",
    "            writer.writerow([chromosome, interested_start, interested_end, gene])\n",
    "            \n",
    "            \n",
    "def qsub_scripts(scripts, wkdir):\n",
    "    \"\"\"qsub scripts\"\"\"\n",
    "    for script in scripts:\n",
    "        print \"cd to %s\" % wkdir\n",
    "        print \"submit %s\" % script\n",
    "        p = subprocess.Popen('ssh tachpc \\\"cd %s;  qsub %s\\\"' %\n",
    "                             (wkdir, script),  shell=True,\n",
    "                             stdout=subprocess.PIPE)\n",
    "        output,  err = p.communicate()\n",
    "\n",
    "        \n",
    "# def make_submit_bedscripts_all_genes(patient_files):\n",
    "#     os.chdir('/home/szong/projects/development/coverage')\n",
    "#     wkdir = os.getcwd()\n",
    "#     #for gene in gene_regions:\n",
    "#     #new_dir =  wkdir + \"/\" + gene\n",
    "#     #os.chdir(new_dir)\n",
    "#     #region_file = gene + \"_combined_exon_pm_1kb.txt\"\n",
    "#     region_file = \"all_genes_combined_exons.txt\"\n",
    "#     out = make_BEDcoverage_script(patient_files, region_file, template_dir)\n",
    "#     BEDcoverage_scripts = out[0]\n",
    "#     complete_stamps = out[1]\n",
    "#     coverage_files = out[2]\n",
    "# #         qsub_scripts(BEDcoverage_scripts, new_dir)\n",
    "# #     qsub_scripts(BEDcoverage_scripts, wkdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "os.chdir('/projects/trans_scratch/validations/workspace/szong/AML_capture/coverage/bedcoverage')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input file has all mandatory columns! Continue...\n"
     ]
    }
   ],
   "source": [
    "start = datetime.datetime.now()\n",
    "meta_file = \"/home/szong/projects/development/coverage/vcf_bam_all_samples_new_20150924.csv\"\n",
    "gene_intervals_file = \"/home/szong/projects/development/coverage/VisCapCancer_input_interval_file.bed\"\n",
    "template_dir = '/home/szong/projects/development/coverage/template/'\n",
    "input_headers = HEADER.INPUT_FILE_HEADER\n",
    "patient_files = get_files(meta_file, input_headers)\n",
    "region_file = \"/home/szong/projects/development/coverage/all_genes_combined_exons.txt\"\n",
    "out = make_BEDcoverage_script(patient_files, region_file, template_dir)\n",
    "BEDcoverage_scripts = out[0]\n",
    "complete_stamps = out[1]\n",
    "coverage_files = out[2]\n",
    "\n",
    "os.chdir('/projects/trans_scratch/validations/workspace/szong/AML_capture/coverage/bedcoverage')\n",
    "wkdir = os.getcwd()\n",
    "\n",
    "qsub_scripts(BEDcoverage_scripts, wkdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_regions = intervals_to_dict(gene_intervals_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "intervals_to_regions(gene_regions)\n",
    "\n",
    "make_submit_bedscripts_all_genes(gene_regions)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
